modin/core/execution/ray/common/deferred_execution.py

```diff
@@ -0,0 +1,707 @@
+    pop = chain.pop
+    tg_e = _Tag.END
+
+    obj = pop()
+    if isinstance(obj, DeferredExecutionException):
+        raise obj
+    if obj is _Tag.REF:
+        obj = refs[pop()]
+    elif obj is _Tag.LIST:
+        obj = []
+        yield _construct_list(obj, chain, refs, meta)
+        if isinstance(obj[0], DeferredExecutionException):
+            raise obj[0]
+
+    while chain:
+        fn = pop()
+        if fn == tg_e:
+            lst.append(obj)
+            break
+
+        if (args_len := pop()) >= 0:
+            if args_len == 0:
+                args = []
+            else:
+                args = chain[-args_len:]
+                del chain[-args_len:]
+                args.reverse()
+        else:
+            args = []
+            yield _construct_list(args, chain, refs, meta)
+        if (args_len := pop()) >= 0:
+            kwargs = {pop(): pop() for _ in range(args_len)}
+        else:
+            values = []
+            yield _construct_list(values, chain, refs, meta)
+            kwargs = {pop(): v for v in values}
+
+        obj = _exec_func(fn, obj, args, kwargs)
+
+        if ref := pop():
+            refs[ref] = obj
+        if pop():
+            if isinstance(obj, ListOrTuple):
+                for o in obj:
+                    meta.append(len(o) if hasattr(o, "__len__") else 0)
+                    meta.append(len(o.columns) if hasattr(o, "columns") else 0)
+                    yield o
+            else:
+                meta.append(len(obj) if hasattr(obj, "__len__") else 0)
+                meta.append(len(obj.columns) if hasattr(obj, "columns") else 0)
+                yield obj
+
+
+def _construct_list(
+    lst: List,
+    chain: List,
+    refs: Dict[int, Any],
+    meta: List,
+):  # pragma: no cover # noqa: GL08
+    pop = chain.pop
+    lst_append = lst.append
+    while True:
+        obj = pop()
+        if isinstance(obj, _Tag):
+            if obj == _Tag.END:
+                break
+            elif obj == _Tag.CHAIN:
+                yield _construct_chain(chain, refs, meta, lst)
+            elif obj == _Tag.LIST:
+                lst_append([])
+                yield _construct_list(lst[-1], chain, refs, meta)
+            elif obj is _Tag.REF:
+                lst_append(refs[pop()])
+            else:
+                raise ValueError(f"Unexpected tag {obj}")
+        else:
+            lst_append(obj)
+
+
+@ray.remote(num_returns=4)
+def remote_exec_func(
+    fn: Callable, obj: Any, *args: Tuple, **kwargs: Dict
+):  # pragma: no cover
+    """
+    Execute the specified function in a worker process.
+
+    The object `obj` is passed to the function as the first argument of.
+
+    Parameters
+    ----------
+    fn : Callable
+    obj : Any
+    *args : list
+    **kwargs : dict
+
+    Returns
+    -------
+    tuple[Any, int, int, str]
+    The execution result, the result length and width, the worked address.
+    """
+    try:
+        obj = _exec_func(fn, obj, args, kwargs)
```

Comment:
We should try serialiazing _exec_func beforehand and pass it as an object ref in to remote_exec_func.

Reply:
I do not see that we serialize `_exec_func` (call ray.put on it) somewhere. It is captured from global scope. I think it might be beneficial to serialize this function beforehand like we do here, for instance. https://github.com/modin-project/modin/blob/5c15c48536486472e8d44eb24772f1c8b49a63f8/modin/core/dataframe/pandas/partitioning/partition.py#L54